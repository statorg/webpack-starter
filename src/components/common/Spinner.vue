<template>
  <div class="spinner" :style="styles">
    <svg
      xmlns="http://www.w3.org/2000/svg"
      :viewBox="viewBox"
      class="spinner__svg"
    >
      <circle
        fill="transparent"
        :cx="viewBoxSize"
        :cy="viewBoxSize"
        :r="radius"
        :stroke-width="strokeWidth"
        :stroke-dasharray="strokeDashArray"
        :stroke-dashoffset="strokeDashOffset"
        class="spinner__circle"
      />
    </svg>
  </div>
</template>

<script>
export default {
  name: 'Spinner',
  props: {
    size: {
      type: [Number, String],
      default: 32
    },
    width: {
      type: [Number, String],
      default: 4
    }
  },
  data () {
    return {
      radius: 20
    }
  },
  computed: {
    styles () {
      return {
        width: `${this.size}px`,
        height: `${this.size}px`
      }
    },
    circumference () {
      return 2 * Math.PI * this.radius
    },
    strokeDashArray () {
      return Math.round(this.circumference * 1000) / 1000
    },
    strokeDashOffset () {
      return `${this.circumference}px`
    },
    strokeWidth () {
      return Number(this.width) / +this.size * this.viewBoxSize * 2
    },
    viewBoxSize () {
      return this.radius / (1 - Number(this.width) / +this.size)
    },
    viewBox () {
      return `0 0 ${this.viewBoxSize * 2} ${this.viewBoxSize * 2}`
    }
  }
}
</script>

<style lang="scss">
.spinner {
  position: relative;
  display: inline-flex;
  vertical-align: middle;
  justify-content: center;
  align-items: center;

  &__svg {
    width: 100%;
    height: 100%;
    margin: auto;
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    z-index: 0;
    animation: progress-circular-rotate 1.4s linear infinite;
    transform-origin: center center;
    transition: all 0.2s ease-in-out;

    @keyframes progress-circular-rotate {
      100% {
        transform: rotate(360deg);
      }
    }
  }

  &__circle {
    stroke: currentColor;
    z-index: 2;
    transition: all 0.6s ease-in-out;
    animation: progress-circular-dash 1.4s ease-in-out infinite;
    stroke-linecap: round;
    stroke-dasharray: 80, 200;
    stroke-dashoffset: 0;

    @keyframes progress-circular-dash {
      0% {
        stroke-dasharray: 1, 200;
        stroke-dashoffset: 0;
      }

      50% {
        stroke-dasharray: 100, 200;
        stroke-dashoffset: -15px;
      }

      100% {
        stroke-dasharray: 100, 200;
        stroke-dashoffset: -125px;
      }
    }
  }
}
</style>
