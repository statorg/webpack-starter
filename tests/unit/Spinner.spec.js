import { shallowMount } from '@vue/test-utils'
import Spinner from '@/components/common/Spinner'

describe('Spinner.vue', () => {
  it('should render and match snapshot', () => {
    const wrapper = shallowMount(Spinner, {
      propsData: {
        size: 23,
        width: 2
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
