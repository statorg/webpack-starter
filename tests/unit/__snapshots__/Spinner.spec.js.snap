// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner.vue should render and match snapshot 1`] = `
<div class="spinner" style="width: 23px; height: 23px;"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 43.80952380952381 43.80952380952381" class="spinner__svg">
    <circle fill="transparent" cx="21.904761904761905" cy="21.904761904761905" r="20" stroke-width="3.8095238095238093" stroke-dasharray="125.664" stroke-dashoffset="125.66370614359172px" class="spinner__circle"></circle>
  </svg></div>
`;
